File: day01/puzzle1.rb
#!/usr/bin/env ruby

# Login to https://adventofcode.com/2025/day/1/input to download 'input.txt'.

# lines = readlines
# lines = File.readlines('sample.txt', chomp: true) # Answer: 3 (in 42 ms)
lines = File.readlines('input.txt', chomp: true) # Answer: 1076 (in 71 ms)

rotations = lines
              .map { |line| line.split(//, 2) }
              .map do |pair|
                case pair[0]
                when 'R' then pair[1].to_i
                when 'L' then -pair[1].to_i
                end
              end

puts 'Rotations'
puts '---------'
puts rotations.inspect
puts

positions = rotations
  .reduce([50]) do |acc, rotation|
    acc << (acc.last + rotation) % 100
    acc
  end

puts 'Positions'
puts '---------'
puts positions.inspect
puts

total = positions.select { |pos| pos == 0 }.size

puts "Total: #{total}"

File: day01/puzzle1.js
#!/usr/bin/env node

import fs from 'node:fs'

const LINE_REGEX = /(?<direction>[LR])(?<distance>\d+)/

// fs.readFile('sample.txt', 'utf8', (err, data) => {
fs.readFile('input.txt', 'utf8', (err, data) => {
    if (err) {
        console.error(err)
        return
    }

    const lines = data.split('\n').filter(s => s)

    console.log('lines')
    console.log(lines)
    console.log()

    const rotations = lines
        .map(line => line.match(LINE_REGEX))
        .map(match => ({ direction: match.groups.direction, distance: Number(match.groups.distance) }))
        .map(({ direction, distance }) => (direction === 'R') ? distance : -distance)

    console.log('rotations')
    console.log(rotations)
    console.log()

    const positions = rotations.reduce((acc, distance) => {
        const previous_position = acc.at(-1)
        let new_position = (previous_position + distance) % 100
        if (new_position < 0) {
            new_position += 100
        }
        acc.push(new_position)
        return acc
    }, [50])

    console.log('positions')
    console.log(positions)
    console.log()

    const total = positions.filter(position => position === 0).length
    console.log(`Total: ${total}`)
})
